### ISSUE TITLE
[Feature Request] Framework for Standardization Tests

### ISSUE THREAD

**Issue by josephineSei** [SovereignCloudStack/standards#359]
For every standard in SCS tests are written. Currently in IaaS these are little scripts from various authors.
It might make programming and testing easier if a framework is used under which all tests on each layer are gathered.

Positive aspects might be:
- easier and faster programming of tests, when work like establishing connections and defining users, etc. is already done
- deployers don't nee to look into how to execute each test, but only, what is necessary for that framework

We should evaluate:

- [ ] does it make sense to have a test framework for IaaS Layer
- [ ] if yes, how should it look like?

**Comment by josephineSei** [comment]
Requirements for an IaaS Test-Framework:
- [ ] one single point of truth for information about the deployment: deployment-connection, test-user, test-credentials, test-project, test-domain, test-domain-manager, etc... (could be a .yaml file or something else)
           -> editable and provided  by the deployer. 
- [ ] responsible for connection to the IaaS with given credentials
- [ ] responsible for providing structure for the cleanup of ressources created in the tests
- [ ] responsible for gathering the results of the tests and present them in some way.
- [ ] maybe (?) sent results to the certificate issuer, which then can issue or revoke certificates
- [ ] maybe (?) parallel execution of tests
- [ ] documentation: how to set up and use this framework (+ maybe how to implement tests)

**Comment by fdobrovolny** [comment]
I went through our current tests, and I have two options as discussed on the last meeting. 

I would say the most value would provide two frameworks:

1. python unittest framework
2. python behave as it is based on the gherkin language

As an example test I selected `k8s-version-recency-check` https://github.com/SovereignCloudStack/standards/blob/main/Tests/kaas/k8s-version-recency/k8s-version-recency-check.py

# Unit test framework

```python
class OpenstackTestCase(unittest.TestCase):
    def setUp(self):
        self.connection = ...

    def test_version-recency(self)
        for cve in self.get_cve():
            with self.subTest(msg=f"CVE {cve.name}"):
                # test CVE

    def tearDown(self):
        self.connection.dispose()
```

# Behave

This kind would enable human readble format of tests.

```
Given a k8s cluster
When we obtain cluster version
Then minor version is less then 4 months old.

Given a k8s cluster
When we obtain cluster version
And we get CVEs
Then no high severity CVE is there within last 8 days.
```

**Comment by josephineSei** [comment]
For CaaS there is already a ticket with a framework, that should be used: https://github.com/SovereignCloudStack/standards/issues/345

That leaves this ticket for IaaS only.

**Comment by josephineSei** [comment]
A possible entry point would be a structure with one or more testsuites (for daily tests, 3-monthly tests, etc..)
It is possible to gather all test in such a structure:
```
--[iaas tests]
   |-flavor-test.py
   |-....-test.py
   |- __init__.py
dailytests.py
monthlytests.py
```
I am currently investigating, if its possible to only provide one connection to the cloud for those testsuites.

**Comment by garloff** [comment]
>    I am currently investigating, if its possible to only provide one connection to the cloud for those testsuites.

Not sure I understand this, sorry to ask.
Like one set of credentials that authorize to talk to many clouds? This does not work.
Like getting a token only once and using it for many tests? But getting a token is cheap (in the order of 1s), so while doing it for every single API call (like you do when using the openstackclient tools from a shell script) does indeed slow down the test, building a connection once per test-case is 1s on a set of API calls that typically take many seconds.
So you must mean something else ... and I don't get it.
